include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -0,0 +1,616 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_MM_CUDA_VM_RESOURCE_H_
+#define DALI_CORE_MM_CUDA_VM_RESOURCE_H_
+
+#include <algorithm>
+#include <mutex>
+#include <utility>
+#include <vector>
+#include "dali/core/mm/cu_vm.h"
+
+#if DALI_USE_CUDA_VM_MAP
+#include "dali/core/mm/memory_resource.h"
+#include "dali/core/mm/detail/free_list.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/bitmask.h"
+#include "dali/core/spinlock.h"
+#include "dali/core/device_guard.h"
+
+namespace dali {
+namespace mm {
+
+namespace test {
+class VMResourceTest;
+}  // namespace test
+
+class cuda_vm_resource : public memory_resource<memory_kind::device> {
+ public:
+  explicit cuda_vm_resource(int device_ordinal = -1,
+                            size_t block_size = 0,
+                            size_t initial_va_size = size_t(4) << 30) {
+    device_ordinal_ = device_ordinal;
+    block_size_ = block_size;
+    initial_va_size_ = initial_va_size;
+
+    configure();
+  }
+  using mem_handle_t = CUmemGenericAllocationHandle;
+
+  ~cuda_vm_resource() {
+    purge();
+  }
+
+  void *try_allocate_from_free(size_t size, size_t alignment) {
+    adjust_params(size, alignment);
+    if (size == 0)
+      return nullptr;
+    lock_guard pool_guard(pool_lock_);
+    // try to get a free region that's already mapped
+    void *ptr = try_get_mapped(size, alignment);
+    if (ptr) {
+      stat_add_allocation(size);
+    }
+    return ptr;
+  }
+
+  void deallocate_no_sync(void *ptr, size_t size, size_t alignment) {
+    deallocate_impl(ptr, size, alignment, false);
+  }
+
+  struct Stat {
+    int allocated_blocks;
+    int peak_allocated_blocks;
+    size_t block_size;
```

Comment:
This field seems to be redundant with just `block_size_`.

Reply:
Yeah, I though about it - I can add a getter for `block_size_` and remove it from `Stat`